rfcs/0009-separate-wal.md

```diff
@@ -0,0 +1,143 @@
+# Separate Object Store for WAL
+
+Status: Draft
+
+Authors:
+
+* [Sergei Sitnikov](https://github.com/taburet)
+
+## Motivation
+
+Sequential writes are a common pattern. Combined with `await_durable` set to
+`true`, which is the default, WAL performance is the main limiting factor for
+the sequential writes throughput. Even if `flush_interval` is set to an extremely
+low value, it can't overcome the inherent latency of the underlying object store.
+
+The obvious solution is switching to a lower latency object store; for instance,
+from S3 to S3 Express One Zone. But lower latency stores usually come with a cost,
+they might be significantly more expensive or might be limited in features. Using
+such stores for the main store containing the entire database might be undesirable.
+
+## Goals
+
+- Support an optional separate object store dedicated specifically for WAL.
+- Provide a user-facing configuration mechanism.
+- Avoid significant refactorings.
+
+## Public API
+
+If user wants to use a separate WAL object store, the store instance should be
+provided while constructing a `Db`:
+
+```rust
+impl Db {
+    // ...
+
+    /// ...
+    ///
+    /// ## Arguments
+    /// ...
+    /// - `wal_object_store`: an optional [object store](ObjectStore) instance
+    ///   dedicated specifically for WAL.
+    ///
+    ///   If set to `None`, the passed `object_store` will be used for WAL storage.
+    ///
+    ///   NOTE: WAL durability and availability properties depend on the properties
+    ///         of the underlying object store. Make sure the configured object
+    ///         store is durable and available enough for your use case.
+    ///
+    /// ...
+
+    pub async fn open_with_opts<P: Into<Path>>(
```

Comment:
I think it makes sense to omit this from the basic `open` method, which realistically is going to be used mostly in demo situations. Just wanted to drop a note though so that the decision is conscious.